### ISSUE TITLE
Overflow error when trying to convert a Python integer to C unsigned long

### ISSUE THREAD

**Issue by fentonscode** [metawards/MetaWards#50]
**Description**
MetaWards fails to complete the tutorial run on the 'lurgy' due to an integer overflow. This is likely a mismatch between 32/64 bit datatypes on Windows systems.

**To Reproduce**
From a fresh install: "metawards --disease lurgy"

Error:

<pre>
...
Random seed for thread 0 equals 3037808250
Using random number seed: 3037808250
Random seed for thread 1 equals 3299740620
Using random number seed: 3299740620
Random seed for thread 2 equals 1715400329
Using random number seed: 1715400329
Random seed for thread 3 equals 1282685305
Using random number seed: 1282685305
Random seed for thread 4 equals 2172416171
Using random number seed: 2172416171
Random seed for thread 5 equals 1505300725
Using random number seed: 1505300725
Random seed for thread 6 equals 1336228674
Using random number seed: 1336228674
Random seed for thread 7 equals 2172219685
Using random number seed: 2172219685
Traceback (most recent call last):
  File "python\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "python\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "Python\Python37\Scripts\metawards.exe\__main__.py", line 7, in <module>
  File "python\python37\lib\site-packages\metawards\app\run.py", line 916, in cli
    profile=profile, parallel_scheme=parallel_scheme)
  File "python\python37\lib\site-packages\metawards\utils\_runner.py", line 136, in run_models
    nthreads=nthreads)
  File "python\python37\lib\site-packages\metawards\_network.py", line 350, in run
    rngs = create_thread_generators(rng, nthreads)
  File "python\python37\lib\site-packages\metawards\utils\_parallel.py", line 138, in create_thread_generators
    return array("L", rngs)
OverflowError: Python int too large to convert to C unsigned long
</pre>

**Environment (please complete the following information):**
 - OS: Windows 10
 - Python: 3.7.4, all optional and development installs from requirements*.txt
 - Version: metawards 0.10.0

**Additional context**
Error is also persistent when either installing via pip or building from source

**Comment by chryswoods** [comment]
Thanks @fentonscode - it's really helpful to have you run this on Windows. I'll sort the fix out now :-)

**Comment by chryswoods** [comment]
The breaking code is here;

https://github.com/metawards/MetaWards/blob/a765acf3950efa8450e82e740c52bd4e4d6dd317/src/metawards/utils/_parallel.py#L139

```python
    # need to return these as an array so that they are more easily
    # accessible from the OpenMP loops - rng is a unsigned 64-bit integer
    # as a uintptr_t - this best corresponds to unsigned long ("L")
    if platform == "win32":
        # Use unsigned long long ("Q") on Windows since it has a 32-bit long.
        return array("Q", rngs)
    else:
        return array("L", rngs)
```

As your code is using the `array("L", rngs)` this suggests that `platform` on your machine is not `win32`. Could you type into Python

```python
import sys
print(sys.platform)
```

and let me know what you get. I suspect it will be something like "win64". If that is the case, then we just need to add that to the if statement so that the `array("Q", rngs)` return is used.

**Comment by chryswoods** [comment]
p.s. I am fixing this in the branch [fix_issue_50](https://github.com/metawards/MetaWards/tree/fix_issue_50). Could you switch to this branch (`git pull` then `git checkout fix_issue_50`) and then I can get you to test some fixes there. Once this is done, we will then pull it into `devel`.

**Comment by fentonscode** [comment]
Python says: <pre>win32</pre>

**Comment by chryswoods** [comment]
I've added some extra print statements to [parallel.py](https://github.com/metawards/MetaWards/blob/b513b36c41a17c67ba5b13fd43673c06e66c1d69/src/metawards/utils/_parallel.py#L136). These will print out the value of `platform` and more detail about which code path is being taken.

Could you pull this into your code and test using that, e.g.

```
git pull
git checkout fix_issue_50
git pull
make
make install

metawards --disease lurgy
```

This will still fail, but should give enough information to enable me to debug.

Thanks :-)

**Comment by fentonscode** [comment]
Will do, I've just put in a PR for a new file I added whilst reading the docs. I'll wait until that's accepted/rejected until I pull the latest version.

**Comment by fentonscode** [comment]
Slightly different error now:

<pre>
Run the model...
Using random number seed: 83266655
First five random numbers equal 47, 50, 62, 42, 48
Number of threads used equals 8
Random seed for thread 0 equals 3472404219
Using random number seed: 3472404219
Random seed for thread 1 equals 644817588
Using random number seed: 644817588
Random seed for thread 2 equals 3394955589
Using random number seed: 3394955589
Random seed for thread 3 equals 1774225101
Using random number seed: 1774225101
Random seed for thread 4 equals 2714231985
Using random number seed: 2714231985
Random seed for thread 5 equals 3898444850
Using random number seed: 3898444850
Random seed for thread 6 equals 2104541659
Using random number seed: 2104541659
Random seed for thread 7 equals 1551943320
Using random number seed: 1551943320
What is the platform?
win32
WINDOWS RETURN Q
Initialise infections...
Traceback (most recent call last):
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\cf502\AppData\Local\Programs\Python\Python37\Scripts\metawards.exe\__main__.py", line 7, in <module>
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\app\run.py", line 978, in cli
    parallel_scheme=parallel_scheme)
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\utils\_run_models.py", line 159, in run_models
    nthreads=nthreads)
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\_network.py", line 489, in run
    iterator=iterator, extractor=extractor)
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\utils\_run_model.py", line 150, in run_model
    profiler=p)
  File "src\metawards\extractors\_output_core.pyx", line 771, in metawards.extractors._output_core.output_core
    output_func(network=network, population=population,
  File "src\metawards\extractors\_output_core.pyx", line 198, in metawards.extractors._output_core.output_core_omp
    def output_core_omp(network: Network, population: Population,
TypeError: output_core_omp() takes exactly 5 positional arguments (4 given)
</pre>

Strangely the *omp() error did not appear on the last compile.

**Comment by lohedges** [comment]
For reference, `sys.platform` always returns `win32` on Windows systems. See [here](https://stackoverflow.com/questions/2144748/is-it-safe-to-use-sys-platform-win32-check-on-64-bit-python) for details. If we need more fine grained control we could use `platinfo`.

For the long sized issue, I read around and it seems that Windows has a 32-bit long for backwards compatibility purposes, i.e. 32-bit applications don't need a recompile. However, on reading further it seems that this might be MSVC++ specific (see, for example, [here](https://stackoverflow.com/questions/384502/what-is-the-bit-size-of-long-on-64-bit-windows)), so perhaps we need to be more careful in general.

The fact that you got to the `Initialise infections...` stage looks like your original build was actually missing the `sys.platform == 'win32'` fix.

**Comment by fentonscode** [comment]
I haven't looked at your C project, but the fixed-width datatypes named *64_t / *32_t might be a sensible idea if they are not already used [See stdint.h]. I had no idea about the workings of `sys.platform`, is it worth using a more native solution baked into a Python extension? (Not sure how good the Python solutions are)

**Comment by chryswoods** [comment]
The python/cython interface makes it difficult to specify data types in this way, as we need to pass the right letter code to the Python array object. The code now works and we'll see if there is an error from the CI.

Looking at your output it seems that you have lots of different metawards installed and they may be clashing with each other.  In your last error you have metawards installed in "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards", while in your original error you have metawards installed in "python\python37\lib\site-packages\metawards"

Are these the same location?

Also, I am confused why you are using a 32 bit python under an emulated environment? Is this normal on Windows? @lohedges - can you give some advice on how you compiled and ran metawards on your machine. Which python are you using.

**Comment by lohedges** [comment]
On Windows I used VSCode with the C++ Desktop Development extension which gave me a C++14 compliant MSVC++ compiler. For Python, I used the latest 64-bit Anaconda for Windows and created a clean Python 3.8 environment. I then built MetaWards within Powershell launched from Anaconda.

**Comment by chryswoods** [comment]
Thanks Lester. @fentonscode - could you give a similar level of detail for your installation. It will be worth documenting as there is something about it which is causing very strange errors.

**Comment by fentonscode** [comment]
@chryswoods: I only have one metawards install, I truncated the previous error messages to reduce the line length and simply forgot in the last paste op. The emulation is native to Windows, any 32 bit app is emulated silently in a x64 installation. However, I feel like having the 64 bit Python might solve this. Mine is 32 bit (some packages I use do not like being run in 64 bit Python). A lot of users might do what I did and simply add metawards to their existing installation and not get a 64 bit Python to run it - so it would appear that either a warning or a workaround put in place would help.

I will report back after adding a 64 bit Python interpreter and having another go.

**Comment by fentonscode** [comment]
Just checked, I have a 64 bit Python installed at the moment, so it has made no difference. I had assumed I was using my (older) 3.6 32 bit one. It gives the same error on a 32 bit 3.7.4 installation.

**Comment by fentonscode** [comment]
Got a bit further

<pre>Run the model...
Using random number seed: 18376521
First five random numbers equal 56, 53, 49, 61, 48
Number of threads used equals 8
Random seed for thread 0 equals 120411857
Using random number seed: 120411857
Random seed for thread 1 equals 2242160647
Using random number seed: 2242160647
Random seed for thread 2 equals 984445374
Using random number seed: 984445374
Random seed for thread 3 equals 1229995868
Using random number seed: 1229995868
Random seed for thread 4 equals 2424316828
Using random number seed: 2424316828
Random seed for thread 5 equals 3847318928
Using random number seed: 3847318928
Random seed for thread 6 equals 1889553011
Using random number seed: 1889553011
Random seed for thread 7 equals 3012421615
Using random number seed: 3012421615
What is the platform?
win32
WINDOWS RETURN Q
Initialise infections...
S: 56082077    E: 0    I: 0    R: 0    IW: 0   TOTAL POPULATION 56082077
S: 56082077  E: 0  I: 0  R: 0  IW: 0  POPULATION: 56082077
Traceback (most recent call last):
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\cf502\AppData\Local\Programs\Python\Python37\Scripts\metawards.exe\__main__.py", line 7, in <module>
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\app\run.py", line 978, in cli
    parallel_scheme=parallel_scheme)
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\utils\_run_models.py", line 159, in run_models
    nthreads=nthreads)
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\_network.py", line 489, in run
    iterator=iterator, extractor=extractor)
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\utils\_run_model.py", line 187, in run_model
    profiler=p2)
  File "src\metawards\iterators\_advance_foi.pyx", line 633, in metawards.iterators._advance_foi.advance_foi
    call_function_on_network(nthreads=nthreads,
  File "c:\users\cf502\appdata\local\programs\python\python37\lib\site-packages\metawards\utils\_get_functions.py", line 292, in call_function_on_network
    **kwargs)
  File "src\metawards\iterators\_advance_foi.pyx", line 187, in metawards.iterators._advance_foi.advance_foi_omp
    cdef unsigned long [::1] rngs_view = rngs
ValueError: Buffer dtype mismatch, expected 'unsigned long' but got 'unsigned long long'
</pre>

Modifications was to add  **kwargs argument to line 772 in _output_core.pyx (the output_func call appears to be missing the thread count when using the parallel version?)

**Comment by chryswoods** [comment]
Thanks - I see what's happened. When I moved you to a fix that branched from 'devel' it introduced you to a bug in the output routine (I am currently working on multi-demographic output so somehow an issue there percolated to devel without triggering a CI error - I will investigate this).

I've pushed a fix to fix_issue_50. Please pull that and it should then work.

**Comment by fentonscode** [comment]
No problem @chryswoods, I usually fix/do a lot more on these before sending them over, but I am extremely limited on how much development work I am allowed to do right now. Otherwise I'd be rolling this all into a single post with fixes. Sorry it's been a long one! Will update my fork and see what happens.

**Comment by chryswoods** [comment]
Yes, you are right, it was missing "profiler=profiler, **kwargs". The kwargs had nthreads added above directly. I'm surprised this passed CI as it should have caused an error in test_integration_ncov.py...

**Comment by fentonscode** [comment]
Error is unchanged (fresh clone as of 1 minute ago)

**Comment by chryswoods** [comment]
Sorry - fixed you problem before last but not the buffer size problem that is now in "src\metawards\iterators\_advance_foi.pyx", line 187". This is an integer size issue again. Looking into this now...

**Comment by chryswoods** [comment]
Sorry for causing you so much trouble @fentonscode - you are suffering the problems of being our first windows user. Yes, the problem is as you said that the size of integer used to move the pointer between Python and C. I've worked out how to get cython to be happy with "uintptr_t" and have changed all "unsigned long"s to "uintptr_t".

I've merged in other changes I've made to "fix_issue_50" and it works now locally for me. Could you pull again and see if this works? Thanks again :-)

**Comment by chryswoods** [comment]
I believe that the pull request I've merged into devel should now fix the issue. If you "git checkout devel" and "git pull" I think the code should compile and run. GitHub closed the issue automatically when I merged the pull request, but please feel free to open this again if it remains a problem.

**Comment by fentonscode** [comment]
No problem. I've definitely done something with the fixed 64 bit types in a Python extension before, although I've never used Cython directly. I still have a close option on my screen, so it appears open despite GH telling me otherwise?

**Comment by chryswoods** [comment]
Sorry - using git too late meant I didn't actually push the fix. I've properly issued the pull request (53) which is running CI now. Once this has verified I will accept it, and then you can test on devel

**Comment by chryswoods** [comment]
It's looking like the CI jobs are queued, so I won't be able to merge this tonight (we're very strict about not merging into devel unless all CI passes).

I'm finishing for the evening as my tiredness level will produce too many mistakes now. If you want to test tonight then pull from feature_demographics, as this is from where I've issued the pull request.

**Comment by chryswoods** [comment]
Tests completed so pull request merged. devel is ready for use.

**Comment by fentonscode** [comment]
Perfect, runs to completion now!

**Comment by chryswoods** [comment]
Excellent - please do let me know if you find any other issues with the tutorial (or need to pull-request in any typos or fixes).
